Add ADF4382

chegbeli opened the pull request:
## Pull Request Description

The ADF4382A is a high performance, ultralow jitter, Frac-N PLL
with integrated VCO ideally suited for LO generation for 5G applications
or data converter clock applications. The high performance
PLL has a figure of merit of -239 dBc/Hz, low 1/f Noise and
high PFD frequency of 625MHz in integer mode that can achieve
ultralow in-band noise and integrated jitter. The ADF4382A can
generate frequencies in a fundamental octave range of 11.5 GHz to
21 GHz, thereby eliminating the need for sub-harmonic filters. The
divide by 2 and 4 output dividers on the part allow frequencies to
be generated from 5.75GHz to 10.5GHz and 2.875GHz to 5.25GHz
respectively.

## PR Type
- [ ] Bug fix (change that fixes an issue)
- [x] New feature (change that adds new functionality)
- [ ] Breaking change (has dependencies in other repos or will cause CI to fail)

## PR Checklist
- [x] I have followed the [Coding style guidelines](http://analogdevicesinc.github.io/no-OS/drivers_guide.html#coding-style)
- [ ] I have performed a self-review of the changes
- [x] I have commented my code, at least hard-to-understand parts
- [ ] I have build all projects affected by the changes in this PR
- [x] I have tested in hardware affected projects, at the relevant boards
- [x] I have signed off all commits from this PR
- [x] I have updated the documentation (wiki pages, ReadMe etc), if applies

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/analogdevicesinc/no-OS?pullRequest=2096) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/analogdevicesinc/no-OS?pullRequest=2096) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/analogdevicesinc/no-OS?pullRequest=2096) it.</sub>

amiclaus reviewed (commented):
some general comments:
- please sign the CLA
- for the documentation build - please add the documentation into the ToC (an example here: https://github.com/analogdevicesinc/no-OS/commit/510041166b2719c1d7182bdefc113db74a5767a9 )
- for the xilinx build make sure that the hardware naming in the build.json file is corresponding to the hdl output name and the project is available on main hdl branch.
- a project README.rst would be nice (see example here: https://github.com/analogdevicesinc/no-OS/commit/6673e2006bb442508840d6d90c5e50860620fca9 )

chegbeli commented:
I'll remove the Xilinx platform as it is not officially supported yet.

amiclaus commented on drivers/frequency/adf4382/adf4382.c:
```diff
@@ -0,0 +1,1199 @@
+ *   @brief  Implementation of adf4382 Driver.
+ *   @author Ciprian Hegbeli (ciprian.hegbeli@analog.com)
+********************************************************************************
+ * Copyright 2023(c) Analog Devices, Inc.
```
2024 :)

amiclaus commented on drivers/frequency/adf4382/adf4382.c:
```diff
@@ -0,0 +1,1199 @@
+#include "no_os_print_log.h"
+#include "no_os_delay.h"
+
+//Charge pump current values expressed in uA
```
nit: c-style comments are preferred even for single line.

amiclaus commented on drivers/frequency/adf4382/adf4382.c:
```diff
@@ -0,0 +1,1199 @@
+
+	ret = no_os_spi_write_and_read(dev->spi_desc, buff,
+				       ADF4382_BUFF_SIZE_BYTES);
+	if(ret < 0)
```
`if (ret)` same as it is done for linux drivers.

amiclaus commented on drivers/frequency/adf4382/adf4382.c:
```diff
@@ -0,0 +1,1199 @@
+
+	*data = buff[2];
+
+	return ret;
```
`return 0;` same as it is done for linux drivers.

amiclaus commented on drivers/frequency/adf4382/adf4382.c:
```diff
@@ -0,0 +1,1199 @@
+	tmp = orig & ~mask;
+	tmp |= data & mask;
+
+	if (tmp != orig)
```
this check is a bit uncommon. but i'm fine with it also.

chegbeli replied:
adapted the function from the Linux implementation https://elixir.bootlin.com/linux/latest/source/drivers/base/regmap/regmap.c#L3137

amiclaus commented on drivers/frequency/adf4382/adf4382.c:
```diff
@@ -0,0 +1,1199 @@
+ * @return	   - 0 in case bits are not set, positive value if they are set
+ * 		     or negative error code otherwise.
+ */
+int adf4382_spi_test_bits(struct adf4382_dev *dev, uint16_t reg_addr,
```
is this function really needed?

rbolboac replied:
the return type should be the execution status of the API. pass the result by pointer in the parameters list

amiclaus commented on drivers/frequency/adf4382/adf4382.c:
```diff
@@ -0,0 +1,1199 @@
+	uint8_t ret;
+	uint16_t i;
+
+	for(i = 0; i < 0x68; i++) {
```
usually we use macro definitions for the registers. but this is probably similar to adf43777 and the register naming is actually the value of the register. looks good like this or macro `AD4382_REG(x)` can be used. Let's see what others prefer.

buha replied:
This is a debugging only function that users wouldn't not use in "production". I wouldn't care much about that. It's probably even easier to follow the datasheet as Cipi wrote it because you normally have a large table with register names and addresses and you see gaps/discontinuities in addresses right away when you look at the addresses.

amiclaus commented on drivers/frequency/adf4382/adf4382.c:
```diff
@@ -0,0 +1,1199 @@
+{
+	int tmp;
+
+	tmp = adf4382_spi_test_bits(dev, 0x20, ADF4382_EN_RDBLR_MSK);
```
is the "test bit" function required here. a getter function should read a register value and obtain the specific bits applying that mask. Here you don't test the bits as function header suggests. you just use the function to read the register value and apply the mask. We have for some older drivers the `read_masked` function implementation but we decided to not use it for newer drivers. so just read and apply mask here. (as in `adf4382_get_ref_div` )

amiclaus commented on drivers/frequency/adf4382/adf4382.c:
```diff
@@ -0,0 +1,1199 @@
+ */
+void adf4382_get_rfout(struct adf4382_dev *dev, uint64_t *val)
+{
+	*val = dev->freq;
```
this should be obtained from hardware.

chegbeli replied:
The value is not readily available from the registers and would need to be recalculated from multiple registers. this would require several multiplication and division operations and would loose precision along the way.

amiclaus commented on drivers/frequency/adf4382/adf4382.c:
```diff
@@ -0,0 +1,1199 @@
+	int enable;
+
+	if(!ch)
+		enable = adf4382_spi_test_bits(dev, 0x2B,
```
same here. use read function + apply mask.

amiclaus commented on drivers/frequency/adf4382/adf4382.c:
```diff
@@ -0,0 +1,1199 @@
+{
+	int enable;
+
+	enable = adf4382_spi_test_bits(dev, 0x2A, ADF4382_PD_SYNC_MSK);
```
ditto.

amiclaus commented on drivers/frequency/adf4382/adf4382.c:
```diff
@@ -0,0 +1,1199 @@
+	uint64_t phase_ci;
+	uint64_t pfd_freq;
+	int ret;
+
```
drop extra line

amiclaus commented on drivers/frequency/adf4382/adf4382.c:
```diff
@@ -0,0 +1,1199 @@
+	if (ret)
+		goto error_spi;
+
+	no_os_udelay(500000);
```
is this value datasheet related? if so, please add appropriate comment. and mdelay can be used instead.

amiclaus commented on drivers/frequency/adf4382/adf4382.c:
```diff
@@ -0,0 +1,1199 @@
+		break;
+	}
+
+	if (vco == 0) {
```
if (vco)

chegbeli replied:
if (!vco)

amiclaus commented on drivers/frequency/adf4382/adf4382.h:
```diff
@@ -0,0 +1,721 @@
+#define ADF4382_SPARE_3D_MSK			NO_OS_BIT(7)
+#define ADF4382_SYNC_SP_DB_MSK			NO_OS_BIT(6)
+#define ADF4382_CMOS_OV_MSK			NO_OS_BIT(5)
+#define ADF4382_CMOS_OV(x)			no_os_field_prep(ADF4382_CMOS_OV_MSK, x)
```
drop this and use field_prep inline.

amiclaus commented on drivers/frequency/adf4382/adf4382.h:
```diff
@@ -0,0 +1,721 @@
+/* ADF4382 REG0020 Map */
+#define ADF4382_EN_AUTOCAL_MSK			NO_OS_BIT(7)
+#define ADF4382_EN_RDBLR_MSK			NO_OS_BIT(6)
+#define ADF4382_EN_RDBLR(x)			no_os_field_prep(ADF4382_EN_RDBLR_MSK, x)
```
drop this, use field_prep inline.

amiclaus commented on drivers/frequency/adf4382/adf4382.h:
```diff
@@ -0,0 +1,721 @@
+#define ADF4382_EN_RDBLR_MSK			NO_OS_BIT(6)
+#define ADF4382_EN_RDBLR(x)			no_os_field_prep(ADF4382_EN_RDBLR_MSK, x)
+#define ADF4382_R_DIV_MSK			NO_OS_GENMASK(5, 0)
+#define ADF4382_R_DIV(x)			no_os_field_prep(ADF4382_R_DIV_MSK, x)
```
ditto.

amiclaus commented on drivers/frequency/adf4382/adf4382.h:
```diff
@@ -0,0 +1,721 @@
+/* ADF4382 REG0054 Map */
+#define ADF4382_ADC_ST_CNV_MSK			NO_OS_BIT(0)
+
+#define ADF4382_SPI_3W_CFG(x)			(no_os_field_prep(ADF4382_SDO_ACTIVE_MSK, x) | \
```
ditto.

amiclaus commented on drivers/frequency/adf4382/adf4382.h:
```diff
@@ -0,0 +1,721 @@
+
+#define KHZ					1000
+#define MHZ					KHZ * KHZ
+#define S_TO_NS					1000000000U
```
same as on linux we have `no_os_units.h`. some macro definitions can be found there. if not, please add them there.

amiclaus commented on drivers/frequency/adf4382/adf4382.h:
```diff
@@ -0,0 +1,721 @@
+	uint8_t				cp_i;
+	uint16_t			bleed_word;
+	uint8_t				ld_count;
+	bool				adf4382a;
```
use device id enum:
https://github.com/analogdevicesinc/no-OS/blob/87b1afc56407fc2f1790e9b1d0c7146002c84d32/drivers/frequency/adf4377/adf4377.h#L606-L610

amiclaus commented on drivers/frequency/adf4382/adf4382.h:
```diff
@@ -0,0 +1,721 @@
+
+/** ADF4382 Remove */
+int adf4382_remove(struct adf4382_dev *dev);
+
```
drop extra line.

amiclaus commented on drivers/frequency/adf4382/iio_adf4382.c:
```diff
@@ -0,0 +1,938 @@
+ *   @brief  Implementation of IIO ADF4382 Driver.
+ *   @author CHegbeli (ciprian.hegbeli@analog.com)
+********************************************************************************
+ * Copyright 2023(c) Analog Devices, Inc.
```
2024

amiclaus commented on drivers/frequency/adf4382/adf4382.h:
```diff
@@ -0,0 +1,721 @@
+ *   @brief  Implementation of adf4382 Driver.
+ *   @author Ciprian Hegbeli (ciprian.hegbeli@analog.com)
+********************************************************************************
+ * Copyright 2023(c) Analog Devices, Inc.
```
2024.

amiclaus commented on drivers/frequency/adf4382/iio_adf4382.c:
```diff
@@ -0,0 +1,938 @@
+		return -EINVAL;
+	adf4382_get_ref_clk(adf4382, &val);
+	return snprintf(buf, len, "%"PRIu64, val);
+//	return 0;
```
drop this.

amiclaus commented on drivers/frequency/adf4382/iio_adf4382.c:
```diff
@@ -0,0 +1,938 @@
+
+	val = en;
+	return iio_format_value(buf, len, IIO_VAL_INT, 1, &val);
+
```
drop extra line.

amiclaus commented on projects/adf4382/Makefile:
```diff
@@ -0,0 +1,15 @@
+# Select the example you want to enable by chossing y for enabling and n for disabling
+BASIC_EXAMPLE = n
+IIO_EXAMPLE = y
+
```
nit: drop extra line

amiclaus commented on projects/adf4382/src/platform/stm32/platform_src.mk:
```diff
@@ -0,0 +1,16 @@
+	$(PLATFORM_DRIVERS)/stm32_irq.c       \
+	$(PLATFORM_DRIVERS)/stm32_gpio_irq.c  \
+	$(PLATFORM_DRIVERS)/stm32_uart.c      \
+	$(PLATFORM_DRIVERS)/stm32_uart_stdio.c
```
add new line.

amiclaus reviewed (commented):
some comments on my side.

rbolboac commented on projects/adf4382/src/platform/xilinx/main.c:
```diff
@@ -0,0 +1,81 @@
+		return ret;
+
+	no_os_uart_stdio(uart_desc);
+	ret = basic_example_main();
```
if (ret) call no_os_uart_remove

rbolboac commented on projects/adf4382/src/platform/stm32/main.c:
```diff
@@ -0,0 +1,79 @@
+		return ret;
+
+	no_os_uart_stdio(uart_desc);
+	ret = basic_example_main();
```
if (ret) call no_os_uart_remove

rbolboac commented on drivers/frequency/adf4382/adf4382.c:
```diff
@@ -0,0 +1,1199 @@
+{
+	uint8_t buff[ADF4382_BUFF_SIZE_BYTES];
+	uint16_t cmd;
+
```
dev should be checked for NULL

rbolboac commented on drivers/frequency/adf4382/adf4382.c:
```diff
@@ -0,0 +1,1199 @@
+	int ret;
+	uint8_t buff[ADF4382_BUFF_SIZE_BYTES];
+	uint16_t cmd;
+
```
dev should be checked for NULL

rbolboac commented on drivers/frequency/adf4382/adf4382.c:
```diff
@@ -0,0 +1,1199 @@
+ */
+int adf4382_set_ref_clk(struct adf4382_dev *dev, uint64_t val)
+{
+	dev->ref_freq_hz = (uint64_t)val;
```
does not look like the cast is needed

rbolboac commented on drivers/frequency/adf4382/adf4382.c:
```diff
@@ -0,0 +1,1199 @@
+ * @param y	- The second value for which to compute.
+ * @return 	- The gcd.
+*/
+static uint32_t greatest_common_divisor(uint32_t x, uint32_t y)
+{
+	uint32_t tmp;
+
+	while (y != 0) {
+		tmp = x % y;
+		x = y;
+		y = tmp;
+	}
+
+	return x;
+}
+
```
should be moved in no_os_util or use the implementation already available there

rbolboac commented on drivers/frequency/adf4382/adf4382.c:
```diff
@@ -0,0 +1,1199 @@
+	pfd_freq = adf4382_pfd_compute(dev);
+	phase_reg_value = no_os_div_u64((phase_ci * pfd_freq), (360 * dev->freq));
+
+	if(phase_reg_value > 255)
+		phase_reg_value -= 255;
```
phase_reg_value maximum value is 255, this will never happen. did you mean == instead of > ?, otherwise please remove lines

rbolboac commented on drivers/frequency/adf4382/adf4382.c:
```diff
@@ -0,0 +1,1199 @@
+	uint8_t pol;
+
+	pol = adf4382_spi_test_bits(dev, 0x32, ADF4382_PHASE_ADJ_POL_MSK);
+	if (pol < 0)
+		return pol;
```
this will never be true, pol is uint8_t, always >= 0, please adapt accordingly

rbolboac commented on drivers/frequency/adf4382/iio_adf4382.c:
```diff
@@ -0,0 +1,938 @@
+				uint32_t *readval)
+{
+	return adf4382_spi_read(dev->adf4382_dev, (uint16_t)reg,
+				(uint8_t *)readval);
```
casting this pointer will result in a different value based on the endianness of the CPU. please do not make the cast

rbolboac commented on drivers/frequency/adf4382/iio_adf4382.c:
```diff
@@ -0,0 +1,938 @@
+				 intptr_t priv)
+{
+	struct adf4382_iio_dev *iio_adf4382 = (struct adf4382_iio_dev *)dev;
+	struct adf4382_dev *adf4382 = iio_adf4382->adf4382_dev;
```
this should be done after the null check of iio_adf4382, please see all occurrences in the file

chegbeli commented:
V1 -> V2
* Removed Xilinx platform, as there is no official support for it
* Updated sdp-ck1z.ioc to use version 6.5.0
* Added missing driver and project documentation
* Fixed wrong POR delay time
* Removed warning messages as they brake the IIO over UART
* Remove adf4382_spi_test_bits and replaced with no_os_field_get
* Added adf4382_get_rfout which computes the current output frequency from register values
* Replaced some MACROs with the ones from Use no_os_units.h
* Removed ADF4382_EN_RDBLR(x), ADF4382_R_DIV(x) and ADF4382_CMOS_OV(x) MACROs
* Added NULL checks to API functions
* Moved dev NULL checks before attributing the dev in the IIO functions
* Fixed pointer casting from u32 to u8
* Added ID structure entries for ADF4382 and ADF4382A
* Added missing no_os_uart_remove to the main files
* Fixed missing adf4382_remove for error conditions in the IIO and basic example

amiclaus reviewed (commented):
there are some conflicts that need addressing.

buha commented on drivers/frequency/adf4382/adf4382.c:
```diff
@@ -0,0 +1,1303 @@
+		return EINVAL;
+	}
+
+	/*Phase adjustment can only be done if bleed is active, and a bleed
```
there is some inconsistency here in how you wrote comments with // or with /*, please make them look alike

buha commented on drivers/frequency/adf4382/adf4382.c:
```diff
@@ -0,0 +1,1303 @@
+		device->clkout_div_reg_val_max = ADF4382A_CLKOUT_DIV_REG_VAL_MAX;
+		break;
+	default:
+		return -EINVAL;
```
you've already allocated something, you'll need to `goto` to a label to avoid memory leaks

buha reviewed (commented):
Looks good, some minor remarks above.
One more thing, Ramona commented about the gcd() function, (like her) I think you should use the util.h one and remove your implementation unless there is a justification for keeping it which would go in a code comment.

chegbeli commented:
> Looks good, some minor remarks above. One more thing, Ramona commented about the gcd() function, (like her) I think you should use the util.h one and remove your implementation unless there is a justification for keeping it which would go in a code comment.

Unfortunately the function in no_os_util.h has a slow execution time which is visible in this driver due to the large numbers which are being processed. I could make a new PR for the gcd replacing the one we currently have, if that is something that would help us.

rbolboac commented on projects/adf4382/src/platform/stm32/platform_src.mk:
```diff
@@ -0,0 +1,17 @@
+	$(PLATFORM_DRIVERS)/stm32_gpio_irq.c  \
+	$(PLATFORM_DRIVERS)/stm32_uart.c      \
+	$(PLATFORM_DRIVERS)/stm32_uart_stdio.c
+	
```
missing new line at the end of file

tfcollins commented:
It doesn't seem like the driver checks for the PLL actually locking. REG 0x58 (bit 0). After initialization it should be the final check.

amiclaus commented on drivers/frequency/adf4382/adf4382.c:
```diff
@@ -0,0 +1,1303 @@
+int adf4382_reg_dump(struct adf4382_dev *dev)
+{
+	uint8_t val;
+	uint8_t ret;
```
if ret can take negative values, then use `int` type. Check throughout the code, since this is occuring multiple times.

amiclaus commented on drivers/frequency/adf4382/adf4382.c:
```diff
@@ -0,0 +1,1303 @@
+
+	ret = no_os_spi_write_and_read(dev->spi_desc, buff,
+				       ADF4382_BUFF_SIZE_BYTES);
+	if(ret)
```
please follow the linux guidelines -> `if (ret)` Check the entire code since there are multiple occurences. Same applies for other statements (`for`, `while`, `switch` etc).

chegbeli commented:
V2 -> V3
*Fixed coding style
*Updated return variables to proper data types
*Add locked check
*Switched to no_os_greatest_common_divisor
*Fixed not freeing memory when failing checks

CiprianRegus commented on drivers/frequency/adf4382/iio_adf4382.c:
```diff
@@ -0,0 +1,989 @@
+ * @return i	- The index of the pair if found or an error if there is no
+ * 		  pair of values in the table.
+ */
+static int adf4382_iio_find_2d_row(const int (*tbl)[2], const int size,
```
You may use `ci_table` directly in this function without needing to pass it as a parameter.
